### ISSUE TITLE
feature/redrawhook breaks zsh-autosuggestions

### ISSUE THREAD

**Issue by frebib** [zsh-users/zsh-syntax-highlighting#579]
Reproducible:

```shell
zsh -f
source path/to/zsh-autosuggestions/zsh-autosuggestions.zsh
source path/to/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh
echo one two three
echo  # notice that 'one two three' are faint from autosuggestions.
```

Repeat as above but with the `feature/redrawhook` branch and the 'autosuggested' text is in fact white instead of the default (? it's gray on mine) gray.

**Comment by danielshahaf** [comment]
Don't have time to investigate right now, but tentatively milestoning 0.7.0 since there's been talk of finally merging redrawhook for that release.

**Comment by danielshahaf** [comment]
Discussion with @romkatv: [users/24616](http://www.zsh.org/cgi-bin/mla/redirect?USERNUMBER=24616)

**Comment by danielshahaf** [comment]
The patches posted in https://github.com/zsh-users/zsh-syntax-highlighting/issues/418#issuecomment-623054593 (one patch to zsh itself, one patch to z-sy-h) seem to fix this issue for me without any change to z-asug.  Could someone confirm, please?  (I don't use z-asug myself.)

**Comment by danielshahaf** [comment]
Since a fix that requires a zsh patch would take a while to reach everyone, we should also figure out a solution — or at least a more graceful failure mode — that doesn't involve changing zsh itself.  What's the best we can do here, without changing zsh itself?

One option is to let z-asug users take the non-redrawhook codepath in z-sy-h, but that would just leave them behind, so it's less than ideal.  It would be better to let everyone use the redrawhook codepath, but make the two plugins interoperable.  @ericfreese Any thoughts about this?

In particular, in light of @romkatv's analysis, would it help if z-asug registered its own pre-redraw hook (when supported by zsh)?  I tried that, but haven't been able to get it to work.  My attempt was:

```zsh
source /usr/share/zsh-autosuggestions/zsh-autosuggestions.zsh
if autoload -U +X add-zle-hook-widget 2>/dev/null; then
        # add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_highlight_reset
        # add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_highlight_apply
        # add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_toggle
        # add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_toggle
        # add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_enable
fi
source /path/to/zsh-syntax-highlighting.zsh # master, with feature/redrawhook merged
```

(I uncommented a different line each time.)

**Comment by danielshahaf** [comment]
There was a subtle bug in feature/redrawhook.  I've fixed it and will push it in a moment (once I have this comment's permalink).  With that bug fixed:

-  Without [the owner=* patch for zsh](https://github.com/zsh-users/zsh/pull/57), if I run

    ```zsh
        autoload add-zle-hook-widget
        add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_fetch
        add-zle-hook-widget zle-line-pre-redraw _zsh_autosuggest_highlight_apply
    ```

    after loading _both_ plugins, then both plugins work.  (If I don't run that, suggestions are displayed not greyed out.)

- With the owner=* patch for zsh, both plugins work regardless of z-asug's calls `add-zle-hook-widget` before or after z-sy-h.

Note that add-zle-hook-widget has an undocumented feature: `add-zle-hook-widget foo 50:bar` can be used to specify the index of a hook.  If z-sy-h and z-asug both pass an index, and z-sy-h passes a larger index, then z-sy-h's hook will run after z-asug's regardless of the order of the add-zle-hook-widget calls.

I'll open a z-asug issue to discuss further.

**Comment by danielshahaf** [comment]
> I'll open a z-asug issue to discuss further.

Here: https://github.com/zsh-users/zsh-autosuggestions/issues/529

**Comment by danielshahaf** [comment]
https://github.com/zsh-users/zsh/pull/57 has been merged to zsh.

With zsh from master (post-merge) and z-sy-h master, I get the autosuggestions after typing the `e` of the last line, but they disappear on typing the `c`.

With zsh from master (post-merge) and z-sy-h feature/redrawhook plus the patch in https://github.com/zsh-users/zsh-syntax-highlighting/issues/418#issuecomment-645928346 to use PR57, this issue is fixed.

I expect this will be fixed by merging redrawhook after making it on by default only on zsh that has the `memo=` feature (https://github.com/zsh-users/zsh/pull/57).

**Comment by danielshahaf** [comment]
Fixed as of cb33cc0081234c0f8e48d89e0ba81cf58a8f7635 — see that commit and its recent parents.  Thanks for the easy-to-read report!
